src/objective-c/GRPCClient/private/GRPCHost.m

```diff
@@ -65,6 +65,28 @@ - (nullable instancetype)initWithAddress:(NSString *)address {
   NSURL *hostURL = [NSURL URLWithString:[@"https://" stringByAppendingString:address]];
   if (hostURL.host && !hostURL.port) {
     address = [hostURL.host stringByAppendingString:@":443"];
+  } else if (!hostURL) {
+    // NSURL failed to parse the address. It probably has a `%` in it (denoting IPv6 zone-id).
+    // Connectivity monitor happens per host.
+    char *chost = NULL, *cport = NULL;
+    if (gpr_split_host_port(address.UTF8String, &chost, &cport)) {
+      return nil;
+    }
+    char *caddr = NULL;
+    if (!cport) {
+      if (gpr_join_host_port(caddr, chost, 443) == -1) {
+        gpr_free(chost);
+        return nil;
+      }
+    } else {
+      caddr = chost;
+    }
+    address = [[NSString alloc] initWithCString:caddr encoding:NSUTF8StringEncoding];
+    if (caddr != chost) {
+      gpr_free(caddr);
+    }
+    gpr_free(chost);
+    gpr_free(cport);
```

Comment:
I did not quite understand this part of code. If the address has a % in it, how are these split and join host/port supposed to help with that?

The solution in my mind was that, since the only thing missing here is `NSURL` not supporting zone id, we will keep the original way of processing the addresses when `NSURL` can recognize them. If NSURL returned `nil`, we do another round of manual check whether the host is a valid ipv6 address with zone id. If that is true, we pass that untouched to core.